ml-fundamentals/ml-experimentation.md

```diff
@@ -0,0 +1,147 @@
+- [Conda](https://docs.conda.io/en/latest/) is a popular package, dependency and environment management framework. It supports multiple stacks (Python, R) and multiple versions of the same environment (e.g. multiple Python versions). `Conda` maintains its own package repository, therefore some packages might not be downloaded and managed directly through `Conda`.
+- [Poetry](https://python-poetry.org/) is a Python dependency management system which manages dependencies in a standard way using `pyproject.toml` files and `lock` files. Similar to `Conda`, `Poetry`'s dependency resolution process is sometimes slow (see [FAQ](https://python-poetry.org/docs/faq/#why-is-the-dependency-resolution-process-slow)), but in cases where dependency issues are common or tricky, it provides a robust way to create reproducible and stable environments.
+
+#### Expected outcomes for virtual environments setup
+
+1. Documentation describing how to create the selected virtual environment and how to install dependencies.
+2. Environment configuration files if applicable (e.g. `requirements.txt` for `venv`, [environment.yml](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#creating-an-environment-from-an-environment-yml-file) for `Conda` or [pyrpoject.toml](https://python-poetry.org/docs/pyproject/) for `Poetry`).
+
+#### Virtual environments benefits
+
+- Productization
+- Collaboration
+- Reproducibility
+
+### Source control and folder/package structure
+
+Applied ML projects often contain source code, notebooks, devops scripts, documentation, scientific resources, datasets and more. We recommend to come up with an agreed folder structure to keep resources tidy. Consider deciding upon a generic folder structure for projects (e.g. which contains the folders `data`, `src`, `docs` and `notebooks`), or adopt popular structures like the [CookieCutter Data Science](https://drivendata.github.io/cookiecutter-data-science/) folder structure.
+
+[Source control](https://github.com/microsoft/code-with-engineering-playbook/tree/master/source-control) should be applied to allow collaboration, versioning, code reviews, traceability and backup. In data science projects, source control should be used for code, and the storing and versioning of other  artifacts (e.g. data, scientific literature) should be decided upon depending on the scenario.
+
+#### Folder structure and source control expected outcomes
+
+- Defined folder structure for all users to use, pushed to the repo.
+- [.gitignore](https://git-scm.com/docs/gitignore) file determining which folders should be synced with `git` and which should be kept locally. For example, [this one](https://github.com/drivendata/cookiecutter-data-science/blob/master/%7B%7B%20cookiecutter.repo_name%20%7D%7D/.gitignore).
+- Determine how notebooks are stored and versioned (e.g. [strip output from Jupyter notebooks](https://github.com/kynan/nbstripout))
```

Comment:
To the point above, we need to balance flexibility with rigor with regard to experiementation in notebooks vs. operationalization. We should have guidance here. Can I create notebooks for quick/dirty experimentation and check them into a private branch I can share? Should I have to pass code quality conventions (linting, unit testing, etc.) to PR that notebook into dev? I think yes to both. Would love to discuss.

Reply:
Good point. Let's have a discussion on more specific guidance.